Clean up and bug fixes

tehhobbit opened the pull request:
Due to updates to the top command in 4.17:

* Fixes a bug where the regex used for ram would cause a stacktrace
* Fixes a bug where the regex for cpu usage would cause a stacktrace

These where older bugs and have never actually worked 

* Fixes a bug where onboard sensors where not displayed
* Fixes a bug where total ram was returned instead of available
* Fixes a bug where only cpu used by userspace was reported

Just to make things a bit easier to read.
* Clean up of overly nested loops

Do you guys want this as one large issue or as several issues ? 

Fixes #84 (adding this so the issue is closed automagically when this is merged)

tehhobbit committed:
Clean up and bug fixes

* Fixes a bug where onboard sensors where not displayed
* Fixes a bug where the regex used for ram would cause a stacktrace
* Fixes a bug where total ram was returned instead of available
* Fixes a bug where the regex for cpu usage would cause a stacktrace
* Fixes a bug where only cpu used by userspace was reported
* Clean up of overly nested loops

mirceaulinic commented:
Hi @tehhobbit,

It's fine fixing them in one single PR. But it would be great to open an issue per bug. Doing so, you can then create a new test case per issue providing the specific mock data, as you see on your device (what triggered the bug in your case). For example: for issue #58 we have provided the mock data under https://github.com/napalm-automation/napalm-eos/tree/develop/test/unit/mocked_data/test_get_bgp_neighbors/issue58_neighbor_down, having the dir name using the issue ID and a very brief description. That way, will be easier on the long term to see what happened, why some lines of code are present etc. Additionally (but not mandatory), if you have new lines, you can also add a short inline comment in your code to point the issue ID.

Does this make sense to you?

Also, make sure pylama is happy (line 476 might be too long - max len is 100 as per https://github.com/napalm-automation/napalm-eos/blob/develop/setup.cfg)

mirceaulinic commented on test/unit/mocked_data/test_get_environment/normal/expected_result.json:
```diff
@@ -351,12 +359,12 @@
   "cpu": {
     "0": {
-      "%usage": 3.5
+      "%usage": 4.3
```
You didn't like the value of `3.5`? :P

tehhobbit replied:
This is an effect of that previously only userspace usage was reported.

mirceaulinic replied:
I was just messing around :)

tehhobbit commented:
For the mock data I am a bit unsure how to do that, since we would have to try the same test with multiple versions of data ( one for older EOS releases and one for newer EOS releases) I have added data to cover the on board sensor case.

mirceaulinic commented:
@tehhobbit  napalm-eos is one of the few napalm drivers (yet) using the new testing framework allowing us to test multiple scenarios, multiple OS versions etc.

Let's look again at the example #58: under `napalm-eos/test/unit/mocked_data/test_get_bgp_neighbors/` (https://github.com/napalm-automation/napalm-eos/tree/develop/test/unit/mocked_data/test_get_bgp_neighbors) you will see two directories:
- normal
- issue58_neighbor_down

All tests have the dir `normal` (for the default behaviour) and anything under must stay unchanged (unless drastic changes such as  executing different commands on the device - as you had in https://github.com/napalm-automation/napalm-eos/pull/76, or other exceptional cases).

Now, for any bug, you only need to do `mkdir <test_case_name>` and place the same files as you see under `normal` but with a different content.

Is this more obvious now?

dbarrosop commented:
Yes, please, add a new test case. Should be fairly easy to add and I think @mirceaulinic is guiding (ask if you have questions). That will allow us to ensure we do regression tests and that we also test corner cases.

Thanks for the fixes, keep them coming ;)

dbarrosop commented:
Yes, please, add a new test case. Should be fairly easy to add and I think @mirceaulinic is guiding (ask if you have questions). That will allow us to ensure we do regression tests and that we also test corner cases.

Thanks for the fixes, keep them coming ;)

dbarrosop commented:
Yes, please, add a new test case. Should be fairly easy to add and I think @mirceaulinic is guiding (ask if you have questions). That will allow us to ensure we do regression tests and that we also test corner cases.

Thanks for the fixes, keep them coming ;)

dbarrosop commented:
Yes, please, add a new test case. Should be fairly easy to add and I think @mirceaulinic is guiding (ask if you have questions). That will allow us to ensure we do regression tests and that we also test corner cases.

Thanks for the fixes, keep them coming ;)

tehhobbit committed:
Tests for newer versions of EOS

dbarrosop commented:
Thx for the last commit, now we are talking. Shouldn't you revert the changes to the "normal" test case, though?

tehhobbit commented:
You mean having the new ones as normal and the exception as it used to be ?

mirceaulinic commented:
As pointed out above, what is under the `normal` dir should remain unchanged, i.e.: exactly the content from https://github.com/napalm-automation/napalm-eos/tree/develop/test/unit/mocked_data/test_get_environment/normal
You only need to add new mock data, not change the existing one in this case.

tehhobbit commented:
Well that is how it is now. So I gather no further changes are needed. The reason for asking is that Arista have changed the output for all future versions making the old behaviour obsolete and for a fact not the normal case.

mirceaulinic commented on test/unit/mocked_data/test_get_environment/normal/show_environment_temperature.json:
```diff
@@ -869,6 +869,33 @@
         }
     ],
     "shutdownOnOverheat": true,
-    "tempSensors": [],
```
@tehhobbit this file was changed - should not be. Please revert this.

mirceaulinic commented on test/unit/mocked_data/test_get_environment/normal/expected_result.json:
```diff
@@ -328,8 +338,6 @@
       "capacity": 2900
     }
   },
-  "used_ram": "",
```
@tehhobbit this file was changed - should not be. Please revert this.

mirceaulinic reviewed (changes requested):
I left few inline comments - please revert the changes from the corresponding files.

dbarrosop commented:
> Well that is how it is now. So I gather no further changes are needed. The reason for asking is that Arista have changed the output for all future versions making the old behaviour obsolete and for a fact not the normal case.

Well, thanks Arista for that : (

However, we can't just adapt the code and break older versions so the patch here has to support both scenarios. Probably a couple of helper methods that are called depending on the version would do.

tehhobbit commented:
Added https://github.com/napalm-automation/napalm-eos/issues/84 for data that is actually wrong (removed in the current PR). In regards to breaking change, nothing of this breaks when I have tested across 4.14 - 4.17 both chassis and fixed formfactor.

dbarrosop commented:
Ok, reviewed the code again. I just realized you just added data to the normal case, you didn't actually changed the test itself. Sorry for the confusion.
